# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [ ]:
# must wrap lambda in parenthesis to self excute 
(lambda num : num + 2)(4)

#### Saving to a Variable

#### Multiple Inputs

#### Passing a Lambda into a Function

In [4]:
def math_operations (f , num):
    return f(num)

math_operations(lambda x : x**2, 10)
math_operations(lambda num : num//10,1000)
my_multiply = lambda num: num
my_multiply_by_ten
math_operations(multiply_by_ten,5)



NameError: name 'multiply_by_ten' is not defined

#### Returning a Lambda from a Function

In [ ]:


def outterFunction():
    otter_num = 4
    def multiply(num):
        inner_num =10
        return num * 4
    return multiply

print(outterFunction()(4))


def returnLambda (num1,num2):
    return lambda num3,num4: num1 + num2 +num3 + num4
#we need to pass the num parameter to run the lambda
add_two_three = returnLambda(2,3)

add_two_three(10,20)

#### If Statements within Lambdas

In [5]:
#if requires an else (only with lambda)
greater_than_10_add = lambda num : num + 2 if num >10 else num
greater_than_10_add(9)


SyntaxError: invalid syntax (1065029834.py, line 2)

In [16]:
def how_to_multiply(num1):
    return lambda num2: num1 * num2
double = how_to_multiply(2)

triple = how_to_multiply(3)

double(10)

triple(10)

30

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [14]:

#Lambda <parameters> : <code to return>

def math_operations (f , num):
    return f(num)
math_operations(lambda num: num **3 , 2)



8

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [20]:
#map(<function>,iterable)
#function should only run 1 value 
def square (num):
    return num**2
number_list = [1,2,3,4,5,6]

#list function
list(map(square,number_list))

[1, 4, 9, 16, 25, 36]

In [22]:
list(map(lambda num : num**3,number_list))
#or use a list compehension 
[num for num in map(lambda num : num**3,number_list)]

[1, 8, 27, 64, 125, 216]

In [23]:
attackers = [5,6,2,8]
defenders = [4,7,1,7]
survivors=list(map(lambda attacker, defender: attacker < defender, attackers, defenders))
print(survivors.count(True))
survivors.count(True) > survivors.count(False)

1


False

#### Using Lambda's with Map

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

In [28]:
nums = [2,3,5,2,3]
result =list(map(lambda num: num*2 - 1,nums))
print(result)
number_list = [1,2,3,4,5,6,7,8,9,10]

#or use a list comprehension 
[num for num in map(lambda num : (num * 2) - 1, number_list)]

[3, 5, 9, 3, 5]


#### Syntax

#### Using Lambda's with Filter()

In [33]:
students = ['hammed','Andy','william','christopher','Will']
list(filter(lambda student:student[0].lower() == 'w',students))

['william', 'Will']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [50]:
import statistics
nums =[3,5,6,6,4,3,6,7,2,1,1,7,5]
(list(filter(statistics.mean(nums)))

SyntaxError: '(' was never closed (666863191.py, line 3)

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [51]:
#must import 
from functools import reduce
#reduce(<function>,iterable)

sum([1,2,3])

def add_two_number (num1, num2):
    return num1+num2 
reduce (add_two_number, [1,2,3,4,5])


15

#### Using Lambda's with Reduce()

In [53]:
def add_num(num):
    if num == 1:
        print(f'add_num({num})=1')
        return 1
    print(f'add_num({num}) = {num} + add_num({num-1})')
    return num+add_num(num-1)

print(add_num(5))

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [52]:
reduce(lambda num1, num2 : num1-num2,[1,2,3,4,5])

-13

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [56]:
#must return a base case 
def add_num (num):
    if num ==1:
        return 1
    print(f'add-num({num}) = {num} + add_num ({num - 1})')
    return num + add_num(num-1)

add_num(5)

add-num(5) = 5 + add_num (4)
add-num(4) = 4 + add_num (3)
add-num(3) = 3 + add_num (2)
add-num(2) = 2 + add_num (1)


15

#### Writing a Factorial Function

In [ ]:
def find_factorial(num):
    if num == 1:
        print(f'find_factorial({num})= 1')
        return num
    print(f'find_factorial{num} = {num} * find_factorial({num - 1})')
    return num * find_factorial(num-1)

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [ ]:
def big_nums(num):
    if num == 1:
        return 1
    return num (big_nums(num-1))
big_nums(4)

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [58]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
def filter_empty_space (place):
    if place.strip() == '':
        return False
    else:
        return True
list(filter(filter_empty_space,places))


['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [17]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

sorted_last_name = lambda names: sorted(names, key=lambda x: x.split()[-1].lower())

print(sorted_last_name(author))


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [72]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

f_temp = list(map(lambda x: (x[0], 9/5 * x[1] + 32), places))

print(f_temp)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [7]:
def fibonacci(fib):
    if fib<2:
        return fib
    else:
        print(f'Iteration: {fib}')
        return (fibonacci(fib-1)+fibonacci(fib-2))
print(fibonacci(5))



Iteration: 5
Iteration: 4
Iteration: 3
Iteration: 2
Iteration: 2
Iteration: 3
Iteration: 2
5
